### ISSUE TITLE
Specify Field order?

### ISSUE THREAD

**Issue by hbooth** [usnistgov/ACVP#92]
Although JSON does not require it, as I am going through implementation I am wondering if we shouldn't require a particular field order. I am trying to be somewhat accepting and will ignore any unknown fields, but there are a few fields that I really need to know in a particular order. Largely, the order they are described in the specs are the order that I need. For the record, I am currently trying to process the input as a stream, if we cannot (or do not) want to specify the order then I will need to use a different implementation (not hard) but the consequence for us will be one of memory usage. I will need to load the whole file in memory and perform some random access. (The library I am using has the ability to create a tree-node representation and I can switch to that.)

A couple of examples:
 - the acv_version field must be first or I don't know how to process the rest of the file.
 - need the algorithm field before any of the other fields and so on

I do need to know sooner rather than later though how we want to address this moving forward.

**Comment by atvassilev** [comment]
I sympathize with the issue you are facing but am leery about enforcing order since JSON does not require it. My gut instinct tells me enforcing an order increases the brittleness of the protocol and this is not a good thing. 

Concerning the memory resource utilization, I guess I don't understand the memory resource problem. My guess is that you will be expanding to hold the entire file in memory anyway at some point, even if you had the fields in the right order and the file is processed as a stream. So, if I my guess is correct, the question is do we expand upfront and then process, or first process a little and expand in the end. 

If the library you are using allows sharing the tree-node representation around all components of the public server that need it, then the memory issue washes out and this becomes a matter of processing flow management. The main visible difference between the two processing alternatives is that you will be able to respond quicker to errors in the version in one case than the other. But this is a small consolation given the problems associated with enforcing an order.

**Comment by hbooth** [comment]
Now we are getting a little more into the weeds, but I am reading in the input and then converting to a "neutral" representation. During this processing I perform all validation and throw away anything I don't need to minimize my internal representation. At the moment, I can likely work around everything but acv_version and keep the stream processing. I do fundamentally need to know what the version of the protocol I should use to process the input. Most other protocol/file formats I know of sticks the version information at the beginning of the format. I could theoretically, read everything into the tree-node representation, validate the tree nodes, and then wrap these nodes with wrapper classes to keep the memory footprint small, but that would complicate a number of things. An alternative would be to convert to tree node, process and validate building the neutral representation with the result of two in-memory representations temporarily. I could also add some caching to process the file until I find acv_version in the stream, but then worst-case a client could stick the acv_version at the end and that wouldn't buy me much.

**Comment by ashman-p** [comment]
Enforcing an order may be tricky. Especially when it comes to optional fields. But in general I am not in-favor.
And isn't the data already in memory as its received from the network?

On Mar 24, 2017, at 4:01 PM, Harold Booth <notifications@github.com<mailto:notifications@github.com>> wrote:


Although JSON does not require it, as I am going through implementation I am wondering if we shouldn't require a particular field order. I am trying to be somewhat accepting and will ignore any unknown fields, but there are a few fields that I really need to know in a particular order. Largely, the order they are described in the specs are the order that I need. For the record, I am currently trying to process the input as a stream, if we cannot (or do not) want to specify the order then I will need to use a different implementation (not hard) but the consequence for us will be one of memory usage. I will need to load the whole file in memory and perform some random access. (The library I am using has the ability to create a tree-node representation and I can switch to that.)

A couple of examples:

  *   the acv_version field must be first or I don't know how to process the rest of the file.
  *   need the algorithm field before any of the other fields and so on

I do need to know sooner rather than later though how we want to address this moving forward.

**Comment by hbooth** [comment]
@ashman-p -- No, the data is not in memory and that is what I am trying to avoid. I am reading off the network and processing the stream as it is being read. I am worried that receiving large amounts of data and loading it into memory will cause issues for the server down the road. I anticipate some of the sub-algorithm specs (i.e. RSA) will result in a fair amount of data.

After some additional analysis I should be able to handle variable order except for the version. I need to know what version of the protocol I am reading in order to process it. Perhaps this means the version needs to be in the http headers instead of in the JSON?

Pretty sure that anyone who needs to support multiple versions of the protocol will be in the same boat if they don't want to read the whole file into memory. If at the end of the day I am in the minority, I can read the whole file into memory, convert it into a tree-node representation and then convert it into a neutral internal representation by processing this tree, but it will be memory-intensive and I feel the protocol will then necessarily require a particular form of implementation.

**Comment by atvassilev** [comment]
We debated adding the protocol version in the http headers in the past and decided against it. The archives may still contain the discussion. In my view, it opens up possibilities for vulnerabilities if you have to rely on incoming http headers for the version. The potential problem is even more severe if we expose the server to the general public to play with it w/o authentication with the no-certificate option.

**Comment by bfussell** [comment]
Something like this ?


POST /validation/acvp/register HTTP1.2
User Agent:
Host: www.acvp.nist.gov
Accept: 
{
"preamble": {
		"acvVersion": "0.3"
	    },
"operation" : "register",
"certificateRequest" : "yes",
"debugRequest" : "no",
"production" : "no",
"encryptAtRest" : "yes",
"oeInformation" : {
.
.
.

**Comment by atvassilev** [comment]
Looks good to me but defer to Harold.

**Comment by hbooth** [comment]
This will work.

**Comment by hbooth** [comment]
Sorry to do this to you, but after some additional thought (and looking at: http://json.org/) would it make more sense to use an array, rather than an object as our top-level data-structure? Arrays are ordered and Objects are unordered. So perhaps:
[
{
  "acvVersion" : "0.3"
},
{
"operation" : "register",
"certificateRequest" : "yes",
"debugRequest" : "no",
"production" : "no",
"encryptAtRest" : "yes",
"oeInformation" : {
.
.
.
}
]

This will allow a specification of an order and still be JSON compliant.

**Comment by hbooth** [comment]
btw, there is nothing to prevent us from just having:

```
[
"0.3",
{
"operation" : "register",
"certificateRequest" : "yes",
"debugRequest" : "no",
"production" : "no",
"encryptAtRest" : "yes",
"oeInformation" : {
.
.
.
}
}
]
```
And just saying the first element is always the version.

**Comment by bfussell** [comment]
No problem, I have not started modifying the docs yet.  Arrays are fine by me, so this would make every message exchange a giant array.  Is that going to impact you pulling only part of the data into memory ?  Is that going to impact omitted fields which is common in JSON ?

**Comment by hbooth** [comment]
I'm only suggesting the use of an array for the top-level, where we either have a "preamble" object or just the version as the first element and then the second element would be the object structure we have developed. I believe an array at the top-level, where order matters, would be superior since that would be in-keeping with the JSON spec.

**Comment by bfussell** [comment]
I'd prefer the keyword included, but I think I need a couple more examples to fully grasp how this would work. Maybe a capabilities response and VsId retry response.   Thanks

**Comment by hbooth** [comment]
The following two require the version as the first value in the array:

Capabilities Response
```
200 OK Response:
[
    "0.2",
    {
        "capabilityResponse": {"vectorSets": [
            {"vsId": 1437},
            {"vsId": 1438}
        ]},
        "testSession": {"testId": 458},
        "accessToken": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiAibmlzdC5vcmciLCJleHAiOiAxNDYyOTg0OTUzLCJjb21wYW55IjogIkNpc2NvIiwianRpIjo0NTh9.zDhd3WrLcb1Xf5-VKOX6k-G70mACbi1tdir8qHAMPyQ"
    }
]
```

VsId retry
```
HTTP 1.2 200 OK
Status: 200 OK
Content-Type: application/json
Content-Length: <length of results>
[
    "0.2",
    {
        "vsId": 1437,
        "retry": 10
    }
]
```

Same examples but with a "preamble" object:
Capabilities Response
```
200 OK Response:
[
    {
        "acv_version": "0.2"
    },
    {
        "capabilityResponse": {"vectorSets": [
            {"vsId": 1437},
            {"vsId": 1438}
        ]},
        "testSession": {"testId": 458},
        "accessToken": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiAibmlzdC5vcmciLCJleHAiOiAxNDYyOTg0OTUzLCJjb21wYW55IjogIkNpc2NvIiwianRpIjo0NTh9.zDhd3WrLcb1Xf5-VKOX6k-G70mACbi1tdir8qHAMPyQ"
    }
]
```

VsId retry
```
HTTP 1.2 200 OK
Status: 200 OK
Content-Type: application/json
Content-Length: <length of results>
[
    {
        "acv_version": "0.2"
    },
    {
        "vsId": 1437,
        "retry": 10
    }
]
```

The advantage of the first is that it is simple and quick. No need to read in an entire object and then find the version field (assuming additional fields are added down the line). If additional fields are necessary, those versions can specify them in the following elements (in other words the second, third or nth elements of the array could vary from version to version).

The disadvantage of the first is that you lose the name/readability and would need to look at the specification to know what this element represents. Given that we make this consistent across the board I don't see this as a huge issue.

The advantages and disadvantages of the second examples are the opposite, you gain readability for people with a tiny bit of fluff. We can also mitigate the processing issues by never adding additional fields to the first object as well (i.e. this first object only ever contains the "acv_version" field). I think it likely boils-down to a style preference. If we were to have lots of elements in the array I would lean toward the second set of examples, but given the version is the only one I would lean (only slightly) toward the first examples.

To be clear I would be fine with either of the above examples.

**Comment by ashman-p** [comment]
It seems like option 2 gives more flexibility in the event of future changes to extend the preamble?

On Mar 27, 2017, at 6:10 PM, Harold Booth <notifications@github.com<mailto:notifications@github.com>> wrote:


The following two require the version as the first value in the array:

Capabilities Response

200 OK Response:
[
    "0.2",
    {
        "capabilityResponse": {"vectorSets": [
            {"vsId": 1437},
            {"vsId": 1438}
        ]},
        "testSession": {"testId": 458},
        "accessToken": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiAibmlzdC5vcmciLCJleHAiOiAxNDYyOTg0OTUzLCJjb21wYW55IjogIkNpc2NvIiwianRpIjo0NTh9.zDhd3WrLcb1Xf5-VKOX6k-G70mACbi1tdir8qHAMPyQ"
    }
]


VsId retry

HTTP 1.2 200 OK
Status: 200 OK
Content-Type: application/json
Content-Length: <length of results>
[
    "0.2",
    {
        "vsId": 1437,
        "retry": 10
    }
]


Same examples but with a "preamble" object:
Capabilities Response

200 OK Response:
[
    {
        "acv_version": "0.2"
    },
    {
        "capabilityResponse": {"vectorSets": [
            {"vsId": 1437},
            {"vsId": 1438}
        ]},
        "testSession": {"testId": 458},
        "accessToken": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiAibmlzdC5vcmciLCJleHAiOiAxNDYyOTg0OTUzLCJjb21wYW55IjogIkNpc2NvIiwianRpIjo0NTh9.zDhd3WrLcb1Xf5-VKOX6k-G70mACbi1tdir8qHAMPyQ"
    }
]


VsId retry

HTTP 1.2 200 OK
Status: 200 OK
Content-Type: application/json
Content-Length: <length of results>
[
    {
        "acv_version": "0.2"
    },
    {
        "vsId": 1437,
        "retry": 10
    }
]


The advantage of the first is that it is simple and quick. No need to read in an entire object and then find the version field (assuming additional fields are added down the line). If additional fields are necessary, those versions can specify them in the following elements (in other words the second, third or nth elements of the array could vary from version to version).

The disadvantage of the first is that you lose the name/readability and would need to look at the specification to know what this element represents. Given that we make this consistent across the board I don't see this as a huge issue.

The advantages and disadvantages of the second examples are the opposite, you gain readability for people with a tiny bit of fluff. We can also mitigate the processing issues by never adding additional fields to the first object as well (i.e. this first object only ever contains the "acv_version" field). I think it likely boils-down to a style preference. If we were to have lots of elements in the array I would lean toward the second set of examples, but given the version is the only one I would lean (only slightly) toward the first examples.

To be clear I would be fine with either of the above examples.

**Comment by atvassilev** [comment]
Guessing these specifications are intended for people and readability is highly desirable, my preference is the second approach.

**Comment by sam1689** [comment]
I also prefer the 2nd approach - but to be specific I believe we
      are using "acvVersion" now instead of "acv_version", correct?
    -Sam
    
    VsId retry
    HTTP 1.2 200 OK
Status: 200 OK
Content-Type: application/json
Content-Length: <length of results>
[
    {
        "acv_version": "0.2"
    },
    {
        "vsId": 1437,
        "retry": 10
    }
]




    
    On 3/28/17 8:13 AM, Apostol Vassilev
      wrote:
    
    
      Guessing these specifications are intended for people and
        readability is highly desirable, my preference is the second
        approach.
      —
        You are receiving this because you are subscribed to this
        thread.
        Reply to this email directly, view
          it on GitHub, or mute
          the thread.
      
        
          
          
        
        
      
      {"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/usnistgov/ACVP","title":"usnistgov/ACVP","subtitle":"GitHub repository","main_image_url":"https://cloud.githubusercontent.com/assets/143418/17495839/a5054eac-5d88-11e6-95fc-7290892c7bb5.png","avatar_image_url":"https://cloud.githubusercontent.com/assets/143418/15842166/7c72db34-2c0b-11e6-9aed-b52498112777.png","action":{"name":"Open in GitHub","url":"https://github.com/usnistgov/ACVP"}},"updates":{"snippets":[{"icon":"PERSON","message":"@atvassilev in #92: Guessing these specifications are intended for people and readability is highly desirable, my preference is the second approach. "}],"action":{"name":"View Issue","url":"https://github.com/usnistgov/ACVP/issues/92#issuecomment-289750280"}}}

**Comment by atvassilev** [comment]
@sam1689 good point, acvVersion is already in the spec so we should preserve it and stay compliant with the style adopted.

**Comment by bfussell** [comment]
Backing up a bit.  Looking at the examples each message is a giant array, even if it is only 2 elements. I'll need to check how disruptive that is to the existing code and get back with you.  I may not have any time prior to Thursday morning to dig into this.

For extensibility and readability, the second approach would be better in my opinion.

**Comment by atvassilev** [comment]
Here is some data that may be relevant for your assessment: http://josh.zeigler.us/technology/web-development/how-big-is-too-big-for-json/

**Comment by ashman-p** [comment]
One quick clarification…
Being new to JSON, I am not sure how to create a stand alone array object (without a keyword).
Given the new proposal show an array [{preamble}, {cap}].

Should it be …
{ [ {preamble}, {cap} ] }

On Mar 27, 2017, at 6:10 PM, Harold Booth <notifications@github.com<mailto:notifications@github.com>> wrote:


The following two require the version as the first value in the array:

Capabilities Response

200 OK Response:
[
    "0.2",
    {
        "capabilityResponse": {"vectorSets": [
            {"vsId": 1437},
            {"vsId": 1438}
        ]},
        "testSession": {"testId": 458},
        "accessToken": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiAibmlzdC5vcmciLCJleHAiOiAxNDYyOTg0OTUzLCJjb21wYW55IjogIkNpc2NvIiwianRpIjo0NTh9.zDhd3WrLcb1Xf5-VKOX6k-G70mACbi1tdir8qHAMPyQ"
    }
]


VsId retry

HTTP 1.2 200 OK
Status: 200 OK
Content-Type: application/json
Content-Length: <length of results>
[
    "0.2",
    {
        "vsId": 1437,
        "retry": 10
    }
]


Same examples but with a "preamble" object:
Capabilities Response

200 OK Response:
[
    {
        "acv_version": "0.2"
    },
    {
        "capabilityResponse": {"vectorSets": [
            {"vsId": 1437},
            {"vsId": 1438}
        ]},
        "testSession": {"testId": 458},
        "accessToken": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiAibmlzdC5vcmciLCJleHAiOiAxNDYyOTg0OTUzLCJjb21wYW55IjogIkNpc2NvIiwianRpIjo0NTh9.zDhd3WrLcb1Xf5-VKOX6k-G70mACbi1tdir8qHAMPyQ"
    }
]


VsId retry

HTTP 1.2 200 OK
Status: 200 OK
Content-Type: application/json
Content-Length: <length of results>
[
    {
        "acv_version": "0.2"
    },
    {
        "vsId": 1437,
        "retry": 10
    }
]


The advantage of the first is that it is simple and quick. No need to read in an entire object and then find the version field (assuming additional fields are added down the line). If additional fields are necessary, those versions can specify them in the following elements (in other words the second, third or nth elements of the array could vary from version to version).

The disadvantage of the first is that you lose the name/readability and would need to look at the specification to know what this element represents. Given that we make this consistent across the board I don't see this as a huge issue.

The advantages and disadvantages of the second examples are the opposite, you gain readability for people with a tiny bit of fluff. We can also mitigate the processing issues by never adding additional fields to the first object as well (i.e. this first object only ever contains the "acv_version" field). I think it likely boils-down to a style preference. If we were to have lots of elements in the array I would lean toward the second set of examples, but given the version is the only one I would lean (only slightly) toward the first examples.

To be clear I would be fine with either of the above examples.

**Comment by atvassilev** [comment]
@ashman-p I am not a JASON expert but it seems you can have a JSON array with no name. Here is an example I found that illustrates how to deal with them: http://stackoverflow.com/questions/24459349/how-to-code-json-array-without-array-name 

Hope this helps...

**Comment by bfussell** [comment]
Norm was gracious enough to dig into the code impact for us and has a change that seems to handle the new format without too much churn.  I'll start modifying the 0.3 specs today.

**Comment by atvassilev** [comment]
That's great, thank you! We will be incorporating this in our server implementation.

**Comment by bfussell** [comment]
I'm going to submit a pull request with only the symmetric cipher spec so we can all agree before I do all the remaining ones.

**Comment by atvassilev** [comment]
Resolved with PR #105.
